File: renderers/android/src/main/kotlin/dev/varn/gui/VarnAccess.kt
package dev.varn.gui

import android.view.View
import android.view.accessibility.AccessibilityNodeInfo
import org.json.JSONObject
import java.util.WeakHashMap

/**
 * What a box says about itself to a reader who cannot see it.
 *
 * A control the platform draws says all of this for itself. One the engine draws is a box with a colour
 * in it, and that is what TalkBack reads unless the tree says otherwise — so a drawn checkbox that says
 * nothing is a checkbox nobody using it can find, let alone tick.
 *
 * What a view says is held beside it rather than on it. A node is rebuilt for every announcement while a
 * prop arrives once, so the three props are gathered into one delegate, and that delegate is found again
 * through a map held only as long as the view is — a view tag is refused for any key that is not an
 * application resource id, and this renderer declares none.
 */
object VarnAccess {
    /** What the platform calls each role, which is the class a control of that kind would have been. */
    private val CLASSES = mapOf(
        "button" to "android.widget.Button",
        "link" to "android.widget.Button",
        "checkbox" to "android.widget.CheckBox",
        "radio" to "android.widget.RadioButton",
        "switch" to "android.widget.Switch",
        "slider" to "android.widget.SeekBar",
        "progressbar" to "android.widget.ProgressBar",
        "image" to "android.widget.ImageView",
        "search" to "android.widget.EditText",
        "tab" to "android.widget.TabWidget",
        "list" to "android.widget.ListView",
        "listitem" to "android.view.View",
        "header" to "android.view.View",
    )

    private val spoken = WeakHashMap<View, Spoken>()

    fun role(view: View, named: String?) {
        held(view).role = named

        view.importantForAccessibility = if (named != null && named != "none") {
            View.IMPORTANT_FOR_ACCESSIBILITY_YES
        } else {
            View.IMPORTANT_FOR_ACCESSIBILITY_AUTO
        }
    }

    fun state(view: View, given: JSONObject?) {
        val about = held(view)

        about.checked = given?.takeIf { it.has("checked") }?.optBoolean("checked")
        about.selected = given?.optBoolean("selected") == true
        view.isEnabled = given?.optBoolean("disabled") != true
    }

    fun reading(view: View, given: JSONObject?) {
        held(view).reading = given
    }

    private fun held(view: View): Spoken {
        spoken[view]?.let { return it }

        val made = Spoken()

        spoken[view] = made
        view.accessibilityDelegate = made

        return made
    }

    class Spoken : View.AccessibilityDelegate() {
        var role: String? = null
        var checked: Boolean? = null
        var selected: Boolean = false
        var reading: JSONObject? = null

        override fun onInitializeAccessibilityNodeInfo(host: View, info: AccessibilityNodeInfo) {
            super.onInitializeAccessibilityNodeInfo(host, info)

            CLASSES[role]?.let { info.className = it }

            checked?.let { ticked ->
                info.isCheckable = true
                info.isChecked = ticked
            }

            info.isSelected = selected

            val given = reading ?: return

            if (given.has("text")) {
                info.stateDescription = given.optString("text")
                return
            }

            if (!given.has("now")) {
                return
            }

            info.rangeInfo = AccessibilityNodeInfo.RangeInfo(
                AccessibilityNodeInfo.RangeInfo.RANGE_TYPE_FLOAT,
                given.optDouble("least", 0.0).toFloat(),
                given.optDouble("most", 1.0).toFloat(),
                given.optDouble("now", 0.0).toFloat(),
            )
        }
    }
}

File: renderers/ios/VarnAccess.swift
import UIKit

/// What a box says about itself to a reader who cannot see it.
///
/// A control the platform draws says all of this for itself. One the engine draws is a box with a colour
/// in it, and that is what a screen reader hears unless the tree says otherwise — so a drawn checkbox
/// that says nothing is a checkbox nobody using VoiceOver can find, let alone tick.
enum VarnAccess {
    /// Answers the traits a role is carried as, since UIKit describes a control by what it behaves like.
    static func traits(_ role: String?) -> UIAccessibilityTraits {
        switch role {
        case "button", "checkbox", "radio", "switch":
            return .button
        case "link":
            return .link
        case "header":
            return .header
        case "image":
            return .image
        case "search":
            return .searchField
        case "slider", "progressbar":
            return .adjustable
        case "tab":
            return [.button, .selected]
        default:
            return .none
        }
    }

    /// Applies what a box is doing, which UIKit says partly in traits and partly in the value.
    ///
    /// There is no trait for being ticked, so a box that is ticked says so the way VoiceOver reads one:
    /// in the value beside the name. A box that cannot be used stops being reachable rather than being
    /// announced and then refusing the touch.
    static func state(_ view: UIView, _ given: [String: Any]?) {
        guard let given else {
            view.accessibilityTraits.remove(.notEnabled)
            view.accessibilityTraits.remove(.selected)
            return
        }

        if let disabled = given["disabled"] as? Bool, disabled {
            view.accessibilityTraits.insert(.notEnabled)
        } else {
            view.accessibilityTraits.remove(.notEnabled)
        }

        if let selected = given["selected"] as? Bool, selected {
            view.accessibilityTraits.insert(.selected)
        } else {
            view.accessibilityTraits.remove(.selected)
        }

        if let checked = given["checked"] as? Bool {
            view.accessibilityValue = checked ? "1" : "0"
        }
    }

    /// Applies what a box that holds a number is at, which is what lets a reader step one by voice.
    static func reading(_ view: UIView, _ given: [String: Any]?) {
        guard let given else {
            return
        }

        if let said = given["text"] as? String {
            view.accessibilityValue = said
            return
        }

        guard let now = VarnValue.number(given["now"]) else {
            return
        }

        let least = VarnValue.number(given["least"])
        let most = VarnValue.number(given["most"])

        if let least, let most, most > least {
            view.accessibilityValue = "\(Int(((now - least) / (most - least)) * 100))%"
            return
        }

        view.accessibilityValue = VarnAccess.said(now)
    }

    /// Answers a number as a reader hears it, which drops a fraction that is not there.
    private static func said(_ value: Double) -> String {
        if value == value.rounded() {
            return String(Int(value))
        }

        return String(value)
    }
}
